### PR TITLE
Add --no-dev option to add and update commands

### PR DESCRIPTION
**Pull request by glensc**
Replaces https://github.com/BrainMaestro/composer-git-hooks/pull/75

### CODE REVIEW

**Comment by glensc** [comment]
pardon for doing some other code style fixes in this PR. if needed, I can move them to separate PR.

noticed some other code smell as well, UpdateCommand extends AddCommand while they leak each other class properties and not all input options are used, shared.

This business logic from AddCommand should be moved to separate class and both classes us that class instead of extending each other.

**Comment by BrainMaestro** [comment]
Hi @glensc this looks great, thanks! After giving it some more thought, I think this should be the default behaviour without any flags. Thoughts? I don't see why the behaviour without `--no-dev` would be desired. Also lets everyone benefit without needing to change any existing scripts.

**Comment by BrainMaestro** [comment]
Also, the changes you made are fine. And I also agree with the code smell between the add and update commands. Might clean that up someday.

**Comment by glensc** [comment]
While I generally agree that nobody likely needs the opposite, I also think it's wise to be in sync with composer model that in default you are in dev mode, and if you need production mode you invoke `--no-dev`.

If needing an option, I think the option should be `--always` as `--dev` seems funny :)

So, your time to roll the ball! ;)

**Comment by BrainMaestro** [comment]
Yes being in sync would be nice, but for us, the production behavior doesn’t make much sense. I’ll go with `always` like you suggested, and it should log a warning that `cghooks` should only be used in dev mode.

**Comment by glensc** [comment]
@BrainMaestro replaced with `--always` option. but please check the logic. I'm really bad inverting boolean logic! :)

File: src/Commands/AddCommand.php [hunk 1]
```diff
@@ -157,4 +166,17 @@ private function setGlobalGitHooksPath()
...
+     * During a composer install or update process,
+     * a variable named COMPOSER_DEV_MODE will be added to the environment.
+     * If the command was run with the --no-dev flag,
+     * this variable will be set to 0, otherwise it will be set to 1.
+     *
+     * @see https://getcomposer.org/doc/articles/scripts.md#defining-scripts
+     */
+    private function isComposerDevMode()
+    {
+        return !$this->always && getenv('COMPOSER_DEV_MODE') === '1';
```
Comment (BrainMaestro) [inline src/Commands/AddCommand.php]: I think we need to split this condition so we can have an error log for when `—always` is used.

**Comment by BrainMaestro** [comment]
@glensc it looks good. One more thing, could we have some test cases? I think we just need two to showcase the behavior of the new always option. One for no dev mode and always, one for dev mode and always.

File: src/Commands/AddCommand.php (hunk 1 above)
Comment (glensc) [reply src/Commands/AddCommand.php]: good idea. moved always check out

**Comment by glensc** [comment]
@BrainMaestro add tests. lots of copy paste there. but did not want to refactor tests in this PR scope.

**Comment by glensc** [comment]
@BrainMaestro can we get this merged and a new release, please?

required for me in https://github.com/eventum/eventum/pull/566#issuecomment-496221548 ;)

**Comment by BrainMaestro** [comment]
@glensc sure, I'll take a look later today. The build is red though.

**Comment by glensc** [comment]
Te tests passed for me locally, I might have pushed other changes that broke. Need to re-check. But if you find the cause, don't hesitate to report ;)

**Comment by BrainMaestro** [comment]
@glensc I found the issue and fixed it. It was just forgetting to unset the variable, so we were stuck in composer dev mode. I also renamed the option from `--always` to `--force-setup`. It fits a bit better with the existing ones. Thanks a lot for all the work you did!

**Comment by BrainMaestro** [comment]
Closes #75

**Comment by glensc** [comment]
@BrainMaestro can you please tag a release, so I can use this?

also, curiosity, why you prefer global functions not methods in a class?

**Comment by BrainMaestro** [comment]
I will create a new release after implementing #79. 

I preferred a free function because the state of the object was not actually being used at all. It's just unfortunate that they have to be global in php

**Comment by glensc** [comment]
you can just create `final` class with `static` methods.

also in php 5.6 you can namespace functions, but I wouldn't recommend that for personal reasons, because I'd actually like to use this in 5.3 projects, and restoring 5.3 compatibilities in this project is not much work. I might submit PR with that sometime :P

**Comment by glensc** [comment]
could you release regardless of https://github.com/BrainMaestro/composer-git-hooks/pull/79

because I see that #79 will take very long to get to usable state, and you can just make another release.

**Comment by BrainMaestro** [comment]
I will work on it myself soon. Won't take much to get it working well. If you're in a rush, you can just use `dev-master` until the release is ready

**Comment by BrainMaestro** [comment]
@glensc new release is ready

**Comment by glensc** [comment]
@BrainMaestro please check why the release is not appearing in packagist: https://packagist.org/packages/brainmaestro/composer-git-hooks

**Comment by BrainMaestro** [comment]
It’s there now.

**Comment by glensc** [comment]
hmm, it still does not work properly. have not yet investigated further:
- https://travis-ci.org/glensc/eventum/jobs/543695325

basically the flow is with this composer.json:
```json

    "scripts": {
        "post-install-cmd": "cghooks add --no-lock",
        "post-update-cmd": "cghooks update"
    },
```

1. `composer install`
2. `composer install --no-dev`


you'll get error for the second command like:
```
> vendor/bin/cghooks add --no-lock
sh: 1: vendor/bin/cghooks: not found
Script vendor/bin/cghooks add --no-lock handling the post-install-cmd event returned with error code 127
```

**Comment by glensc** [comment]
so, seems we still need to handle this in shell level? i.e as done in original PR:
- https://github.com/BrainMaestro/composer-git-hooks/pull/75#pullrequestreview-242006534

**Comment by glensc** [comment]
or do as class events? because there's a similar problem with other event handler, and those are just skipped, but not made as fatal:

```
Process class 'ComponentInstaller\Process\CopyProcess' not found, skipping this process
Process class 'ComponentInstaller\Process\RequireJsProcess' not found, skipping this process
Process class 'ComponentInstaller\Process\RequireCssProcess' not found, skipping this process
Process class 'ComponentInstaller\Process\BuildJsProcess' not found, skipping this process
```

**Comment by BrainMaestro** [comment]
How would that look like?

**Comment by glensc** [comment]
what is "that"? be specific if replying to three notes at once.

**Comment by BrainMaestro** [comment]
I was asking about the class events

**Comment by glensc** [comment]
here's the docs: https://getcomposer.org/doc/articles/scripts.md#event-classes

**Comment by BrainMaestro** [comment]
Yeah that might be doable. Can you create an issue?

**Comment by glensc** [comment]
super dumb that you reverted everything, including commits that improved other areas.

**Comment by BrainMaestro** [comment]
@glensc I'm sure you can express displeasure without being insulting. The reason everything was reverted is because the commits were squashed (there were 17 of them) and that was the easiest path. I haven't had the time since, or I would have cherry-picked the other relevant commits that improved other stuff. I will still get around to doing that, as the improvements were nice

**Comment by glensc** [comment]
moved: https://github.com/BrainMaestro/composer-git-hooks/pull/95

my commits are [atomic](https://www.freshconsulting.com/atomic-commits/) for the same purpose to revert the specific change. you shouldn't squash them to a single one.

**Pull request merged**